sqlglot/optimizer/scope.py

```diff
@@ -511,82 +535,99 @@ def _traverse_union(scope):
+            union = cte.this
 
             if isinstance(union, exp.Union):
                 recursive_scope = scope.branch(union.this, scope_type=ScopeType.CTE)
 
         for child_scope in _traverse_scope(
             scope.branch(
-                derived_table if isinstance(derived_table, exp.UDTF) else derived_table.this,
-                chain_sources=sources if scope_type == ScopeType.CTE else None,
-                outer_column_list=derived_table.alias_column_names,
-                scope_type=ScopeType.UDTF if isinstance(derived_table, exp.UDTF) else scope_type,
+                cte.this,
+                chain_sources=sources,
+                outer_column_list=cte.alias_column_names,
+                scope_type=ScopeType.CTE,
             )
         ):
             yield child_scope
 
-            # Tables without aliases will be set as ""
-            # This shouldn't be a problem once qualify_columns runs, as it adds aliases on everything.
-            # Until then, this means that only a single, unaliased derived table is allowed (rather,
-            # the latest one wins.
-            alias = derived_table.alias
+            alias = cte.alias
             sources[alias] = child_scope
 
             if recursive_scope:
                 child_scope.add_source(alias, recursive_scope)
 
         # append the final child_scope yielded
-        if is_cte:
-            scope.cte_scopes.append(child_scope)
-        else:
-            scope.derived_table_scopes.append(child_scope)
+        scope.cte_scopes.append(child_scope)
 
     scope.sources.update(sources)
 
 
-def _add_table_sources(scope):
+def _traverse_tables(scope):
     sources = {}
-    for table in scope.tables:
-        table_name = table.name
 
-        if table.alias:
-            source_name = table.alias
-        else:
-            source_name = table_name
+    # Traverse FROMs, JOINs, and LATERALs in the order they are defined
+    expressions = []
+    from_ = scope.expression.args.get("from")
+    if from_:
+        expressions.extend(from_.expressions)
 
-        if table_name in scope.sources:
-            # This is a reference to a parent source (e.g. a CTE), not an actual table.
-            scope.sources[source_name] = scope.sources[table_name]
+    for join in scope.expression.args.get("joins") or []:
+        expressions.append(join.this)
+
+    for lateral in scope.expression.args.get("laterals") or []:
+        expressions.append(lateral)
+
+    for expression in expressions:
+        if isinstance(expression, exp.Table):
+            table_name = expression.name
+            source_name = expression.alias_or_name
+
+            if table_name in scope.sources:
+                # This is a reference to a parent source (e.g. a CTE), not an actual table.
+                sources[source_name] = scope.sources[table_name]
+            else:
+                sources[source_name] = expression
+            continue
```

Comment:
we can also do if-elif-else here to remove this `continue`

Reply:
It'd have to be
```python
if ...:
    ...
else:
    if ...:
        ...
    else:
        ...
    for child_scope in _traverse_scope(...):
        ...
```

Reply:
Oh sorry, I missed the following `for` block. Makes sense.